### PR TITLE
[DSpark] Support pipeline-parallel targets in aggregated serving

### PR DESCRIPTION
**Pull request by lucifer1004**
## What

DSpark on a pipeline-parallel target in aggregated (non-PD) serving. The drafter runs wholly on the last PP stage; this PR makes that work correctly:

- **Draft broadcast across stages**: the last stage's speculator proposes drafts, but earlier stages need them for the next verification step. The last rank now broadcasts the fresh drafts on the PP group (`PPHandler.broadcast_drafts`), and the broadcast is gated out of the no-op second call site when a speculator ran (double-posting misaligned the recv FIFO and hung the pipeline).
- **Warmup**: the deferred PP post-update triton kernel never runs during warmup steps on non-last ranks; its first mid-serving compile would deadlock against the in-flight NCCL broadcast. `warmup_pp_decode_update` compiles it during capture, and the sampled-token broadcast is disabled for the warmup window.
- **Draft embedding under PP**: the target's embedding table only exists on the first stage, so DSv4/K3 DSpark drafters load their own copy from the checkpoint (`loads_own_embed_under_pp`) instead of aliasing.
- **Padded graph batch safety**: the DFlash prepare-inputs kernel clears `input_ids`/`positions` and sets `is_padding` on CUDA-graph padding rows, and the DSv4 top-k router zeroes padded-row selections instead of reading uninitialized state.

## Not a duplicate

Split from #53577 (reviewer request): this PR is the aggregated-serving half; the PD prefill + KV-transfer half follows separately.

## Tests

- New/extended: `tests/v1/worker/test_pp_utils.py`, `test_spec_decode_embed_sharing_pp.py`, `tests/v1/spec_decode/test_dflash_prepare_inputs.py`, `tests/models/kimi_k3/test_eagle3.py`, `test_dspark_mla.py`, `tests/kernels/moe/test_topk_softplus_sqrt.py`.
- `pytest tests/v1/worker/test_pp_utils.py tests/v1/worker/test_spec_decode_embed_sharing_pp.py tests/models/kimi_k3/test_eagle3.py tests/models/kimi_k3/test_dspark_mla.py tests/v1/spec_decode/test_dflash_prepare_inputs.py tests/v1/worker/test_gpu_warmup_blocks.py tests/v1/worker/test_gpu_model_runner_v2.py tests/kernels/moe/test_topk_softplus_sqrt.py -q` → 1744 passed.

## Model evaluation

DeepSeek-V4-Flash (0731) DSpark K=5, IFB PP2×TP2, GSM8K strict-match: **0.9545** (matches the target-only arm within noise).

Kimi-K3 DSpark IFB PP additionally needs #51065 (without it, the draft's non-causal flag raises the shared KV group's TritonMLA reorder threshold and misroutes the causal target's verification blocks).

AI assistance was used in preparing this PR.

### CODE REVIEW

File: vllm/models/kimi_k3/nvidia/dspark_mla.py [hunk 1]
```diff
@@ -141,8 +154,18 @@ def __init__(
...
+        # first stage while the drafter runs on the last, so the draft builds
+        # its own table and loads embed_tokens.weight from its checkpoint
+        # (the K3 DSpark checkpoint always ships it).
         self.embed_tokens: nn.Module | None = None
+        if _target_pp_world_size() > 1:
+            self.embed_tokens = VocabParallelEmbedding(
+                self.config.vocab_size,
+                self.config.hidden_size,
+                prefix=maybe_prefix(prefix, "embed_tokens"),
+            )
```
Comment (GirasoleY) [inline vllm/models/kimi_k3/nvidia/dspark_mla.py:168]: https://github.com/vllm-project/vllm/pull/50514 added token embedding at pp last stage, I think we don't need this anymore (same for the dsv4 path)

File: vllm/v1/worker/gpu_worker.py [hunk 2]
```diff
@@ -771,6 +771,15 @@ def initialize_from_config(self, kv_cache_config: KVCacheConfig) -> None:
...
     @instrument(span_name="Warmup (GPU)")
     def compile_or_warm_up_model(self) -> CompilationTimes:
+        # All warmup phases below run synthetic steps whose sampled outputs are
+        # discarded. The PP sampled-token broadcast would carry no payload, and
+        # its side-stream NCCL ops can overlap the next step's activation p2p
+        # and deadlock the pipeline, so keep it disabled for the whole warmup
+        # window and restore it before serving.
+        pp_handler = getattr(self.model_runner, "pp_handler", None)
+        if pp_handler is not None:
+            pp_handler.set_disabled(True)
```
Comment (GirasoleY) [inline vllm/v1/worker/gpu_worker.py:781]: Could you help me understand how the deadlock happen? If it's about collective running on multiple streams, I'd prefer enforce collective ordering instead of skip warm up here.

Comment (lucifer1004) [reply vllm/v1/worker/gpu_worker.py:781]: Happy to walk through it. The deadlock chain we observed (5-11 min hangs, reproducible):

1. The sampled-token broadcast is posted on a side stream: non-last ranks post NCCL recv kernels that spin on-device until the last rank's broadcast lands (early posting is by design for pipelining).
2. Warmup runs synthetic steps whose sampled outputs are discarded — those steps post the recv spins but carry no matching payload, so the spins linger.
3. Warmup is also where most first-time Triton JIT loads happen. Even on a cubin cache hit, a new specialization key (alignment/divisibility) lazily triggers a fresh cuModuleLoadData — and that call blocks at context level behind the spinning NCCL kernel (driver behavior we can't change from user space).
4. The blocked host thread never advances to the next step, so it never sends its activation p2p; the peer stage waits on those activations; the spinning recv never completes. Cycle → pipeline hang.

On enforcing collective ordering instead: ordering the broadcast against the compute stream would reintroduce a per-step serialization cost in *serving* to fix a warmup-only hazard, and it still wouldn't cover the module-load half of the cycle (a new shape specialization mid-serving would wedge behind a legitimately in-flight broadcast — that's why we precompile the deferred post-update kernel during capture via warmup_pp_decode_update, with alignment-matched views). The disable window here is scoped to synthetic warmup steps that carry no payload by construction, so serving behavior is untouched. If you still see a cleaner ordering-based formulation I'm happy to try it.

File: vllm/models/kimi_k3/nvidia/dspark_mla.py (hunk 1 above)
Comment (lucifer1004) [reply vllm/models/kimi_k3/nvidia/dspark_mla.py]: Confirmed — removed in c6a93432c6. `loads_own_embed_under_pp` and the `load_own_embed` block are gone from both the K3 (`dspark_mla.py`) and DSv4 (`dspark.py`) drafters; the draft now relies on #50514's token embedding provided at the last PP stage.

Verification on this branch:
- Unit: the 60 dspark/PP-related tests in `tests/config/test_dspark_prefill_only.py`, `tests/v1/worker/test_gpu_model_runner_v2.py` and `tests/v1/worker/test_gpu_warmup_blocks.py` all pass.
- e2e (IFB, aggregated serving): DeepSeek-V4-Flash DSpark pp2tp2 GSM8K = 0.9484 (strict-match, 1319 samples; vs 0.9545 measured on this branch before the removal — within noise), spec decode mean acceptance length 4.38.
- Kimi-K3 DSpark pp2tp4 smoke (with #51065): coherent outputs across math/Chinese/code prompts and healthy per-position acceptance (144/96/55/31/22/17/12/6). Note: with #50514 the embedding is shared from the target, so the drafter also saves one embedding copy of memory versus the old self-loading path.

**Review by GirasoleY — APPROVED** [review]
The functionality is correct. Thanks for contributing :)

I opened a PR to change the implementation a bit so we do not need to disable/enable state during capture. Feel free to review and merge it.

**Review by zyongye — APPROVED** [review]

**Comment by njhill** [comment]
@lucifer1004 apologies we reverted this because it needs some more streamlining. We're very sensitive to adding complexity to the core model runner files. Please re-open and I can help iterate some more on it!
